typed-racket-lib/typed-racket/base-env/prims-struct.rkt

```diff
@@ -99,13 +109,25 @@
 
 (define-syntax (define-typed-struct/exec stx)
   (syntax-parse stx #:literals (:)
-    [(_ nm ((~describe "field specification" [fld:optionally-annotated-name]) ...) [proc : proc-ty])
+    [(_ nm:struct-name ((~describe "field specification" [fld:optionally-annotated-name]) ...)
+        [proc : proc-ty] (~optional (~seq #:type-name type:id)))
      (with-syntax*
-      ([proc* (with-type* #'proc #'proc-ty)]
+      ([type (or (attribute type) #'nm.name)]
+       [proc* (with-type* #'proc #'proc-ty)]
        [d-s (ignore-some (syntax/loc stx (define-struct nm (fld.name ...)
                                       #:property prop:procedure proc*)))]
-       [dtsi (quasisyntax/loc stx (dtsi/exec* () nm (fld ...) proc-ty))])
-      #'(begin d-s dtsi))]))
+       [stx-err-fun (if (not (free-identifier=? #'nm.name #'type))
+                        (syntax/loc stx
+                          (define-syntax type
+                            (lambda (stx)
+                              (raise-syntax-error
+                               'type-check
+                               "type name used out of context"
```

Comment:
Can we say what type name this is in the error message, along the lines of @jsmaniac's recent change?